### PR TITLE
Remove Parameterized

### PR DESCRIPTION
**Pull request by nish21**
This PR removes Parameterized from the codebase. 
Particularly, this PR 
- removes `garage.core.parameterized`
- removes `garage.tf.core.parameterized`
- replaces all base classes inheriting from Parameterized with the new base classes used by the models primitive. 

Note:
I had to remove a test for the `Serializable` class as that depended on some functions from `Parameterized`. So removal of `Paramtereized` would be blocked by that one test. Since we are in the process of removing `Serializable` as well, I figured it is okay to remove this test.

### CODE REVIEW

File: src/garage/np/policies/base.py [hunk 1]
```diff
@@ -1,68 +1,78 @@
-from garage.core import Parameterized
+"""Base class for policies based on numpy."""
 
 
-class Policy(Parameterized):
+class Policy():
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: remove ()

Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please make this an ABC while you're here.

File: src/garage/np/policies/base.py [hunk 2]
```diff
@@ -1,68 +1,78 @@
...
 
     def terminate(self):
-        """
-        Clean up operation
-        """
+        """Clean up operation."""
         pass
 
 
 class StochasticPolicy(Policy):
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: also an ABC

File: src/garage/np/policies/base.py (hunk 1 above)
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please update these docstrings while you're here.

File: src/garage/tf/policies/base.py [hunk 3]
```diff
@@ -1,25 +1,49 @@
...
+"""Policy base classes without Parameterized."""
+import tensorflow as tf
 
+from garage.misc.tensor_utils import flatten_tensors, unflatten_tensors
 
-class Policy(Parameterized):
-    def __init__(self, env_spec):
-        Parameterized.__init__(self)
+
+class Policy:
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: please make this an ABC

File: src/garage/tf/policies/base.py [hunk 4]
```diff
@@ -1,25 +1,49 @@
...
+        self._name = name
         self._env_spec = env_spec
+        self._variable_scope = None
+        self._cached_params = {}
+        self._cached_param_shapes = {}
 
     # Should be implemented by all policies
-
     def get_action(self, observation):
+        """Get action given observation."""
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: please make these docstrings fully-written while you're here

File: src/garage/tf/policies/base.py [hunk 5]
```diff
@@ -64,45 +88,105 @@ def state_info_keys(self):
...
+        del new_dict['_cached_params']
+        return new_dict
+
+    def __setstate__(self, state):
+        """Object.__setstate__."""
+        self._cached_params = {}
+        self.__dict__.update(state)
+
 
 class StochasticPolicy(Policy):
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: also an ABC

File: src/garage/tf/regressors/base.py [hunk 6]
```diff
@@ -1,33 +1,37 @@
...
+class Regressor(abc.ABC):
     """
-    Regressor base class.
+    Regressor base class without Parameterized.
 
     Args:
-        input_shape: Input shape.
-        output_dim: Output dimension.
-        name: Name of the regressor.
+        input_shape (tuple): Input shape.
```
Comment (ryanjulian) [inline src/garage/tf/regressors/base.py]: when using containers, please specify the type of the contained objects (i.e. `tuple[int]`)

File: src/garage/tf/regressors/base.py [hunk 7]
```diff
@@ -36,23 +40,83 @@ def predict(self, xs):
...
+
+        Return:
+            The predicted ys.
         """
         raise NotImplementedError
 
     def get_params_internal(self, **tags):
-        """Get the list of parameters."""
-        return self._variable_scope.trainable_variables()
+        """
```
Comment (ryanjulian) [inline src/garage/tf/regressors/base.py]: this needs a return type

File: src/garage/np/policies/base.py [hunk 8]
```diff
@@ -1,68 +1,88 @@
...
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
+        Returns:
+            action (np.ndarray): Action sampled from the policy.
+
+        """
         raise NotImplementedError
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: `pass`

File: src/garage/np/policies/base.py [hunk 9]
```diff
@@ -1,68 +1,88 @@
-from garage.core import Parameterized
+"""Base class for policies based on numpy."""
+import abc
 
 
-class Policy(Parameterized):
+class Policy(abc.ABC):
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please document constructor params here

File: src/garage/np/policies/base.py [hunk 10]
```diff
@@ -1,68 +1,88 @@
...
+        """
         raise NotImplementedError
 
     def reset(self):
+        """Reset the policy."""
         pass
 
     @property
     def observation_space(self):
+        """Return the observation space of the environment."""
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: properties don't need the verb "Return"

Comment (ryanjulian) [reply src/garage/np/policies/base.py]: properties also need types

File: src/garage/np/policies/base.py [hunk 11]
```diff
@@ -1,68 +1,88 @@
...
-
+    @abc.abstractmethod
     def get_action(self, observation):
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
+        Returns:
+            action (np.ndarray): Action sampled from the policy.
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: return documentation doesn't include a name.

see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for an example

File: src/garage/np/policies/base.py (hunk 10 above)
Comment (ryanjulian) [reply src/garage/np/policies/base.py]: see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for examples

File: src/garage/np/policies/base.py [hunk 12]
```diff
@@ -1,68 +1,88 @@
...
     @property
     def action_space(self):
+        """Return the action space for the environment."""
         return self._env_spec.action_space
 
     @property
     def recurrent(self):
         """
-        Indicates whether the policy is recurrent.
+        Indicate whether the policy is recurrent.
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please remove docstring markup like `:return:` -- this is from a different docstring style.

File: src/garage/np/policies/base.py [hunk 13]
```diff
@@ -1,68 +1,88 @@
...
 
     @property
     def state_info_keys(self):
         """
-        Return keys for the information related to the policy's state when
-        taking an action.
+        Get keys describing policy's state.
+
+        Return keys for the information related to the policy's state
+        when taking an action.
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: needs a return type

File: src/garage/np/policies/base.py (hunk 8 above)
Comment (ryanjulian) [reply src/garage/np/policies/base.py]: please use pass rather than NotImplementedError

File: src/garage/np/policies/base.py [hunk 14]
```diff
@@ -1,68 +1,88 @@
...
+
         :rtype Distribution
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
     def dist_info_sym(self, obs_var, state_info_vars):
         """
         Return the symbolic distribution information about the actions.
+
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please remove these old `:param:` style documentations and replace them with the Google style.

File: src/garage/tf/policies/base.py [hunk 15]
```diff
@@ -1,25 +1,69 @@
-from garage.tf.core import Parameterized
+"""Policy base classes without Parameterized."""
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: Base class for Policies

File: src/garage/tf/policies/base.py [hunk 16]
```diff
@@ -1,25 +1,69 @@
...
     # Should be implemented by all policies
-
+    @abc.abstractmethod
     def get_action(self, observation):
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
+        Returns:
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: returns don't have names.

File: src/garage/tf/policies/base.py [hunk 17]
```diff
@@ -1,25 +1,69 @@
...
+            action (np.ndarray): Action sampled from the policy.
+
+        """
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: please use `pass` rather than `NotImplementedError`

File: src/garage/tf/policies/base.py [hunk 18]
```diff
@@ -1,25 +1,69 @@
...
     def get_actions(self, observations):
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observations (list[np.ndarray]): Observations from the environment.
+        Returns:
+            actions (np.ndarray): Actions sampled from the policy.
+
+        """
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: pass

File: src/garage/tf/policies/base.py [hunk 19]
```diff
@@ -1,25 +1,69 @@
...
         raise NotImplementedError
 
     def reset(self, dones=None):
+        """Reset policy."""
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: needs documentation for dones

File: src/garage/tf/policies/base.py [hunk 20]
```diff
@@ -64,45 +108,133 @@ def state_info_keys(self):
...
+    def get_params(self, trainable=True):
+        """Get the trainable variables."""
+        return self.get_trainable_vars()
+
+    def get_param_shapes(self, **tags):
+        """Get parameter shapes."""
+        tag_tuple = tuple(sorted(list(tags.items()), key=lambda x: x[0]))
+        if tag_tuple not in self._cached_param_shapes:
+            params = self.get_params(**tags)
+            param_values = tf.get_default_session().run(params)
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: the CI complains that you need to use `tf.compat.v1.get_default_session`

File: src/garage/tf/policies/base.py [hunk 21]
```diff
@@ -64,45 +108,133 @@ def state_info_keys(self):
...
+        Args:
+            tags (dict): A map of parameters for which the values are required.
+        Returns:
+            param_values (np.ndarray): Values of the parameters evaluated in
+            the current session
+
         """
-        pass
+        params = self.get_params(**tags)
+        param_values = tf.get_default_session().run(params)
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: `tf.compat.v1.get_default_session`

File: src/garage/tf/regressors/base.py [hunk 22]
```diff
@@ -36,23 +40,97 @@ def predict(self, xs):
...
+        if tag_tuple not in self._cached_params:
+            self._cached_params[tag_tuple] = self.get_params_internal(**tags)
+        return self._cached_params[tag_tuple]
+
+    def get_param_shapes(self, **tags):
+        """Get the list of shapes for the parameters."""
+        tag_tuple = tuple(sorted(list(tags.items()), key=lambda x: x[0]))
+        if tag_tuple not in self._cached_param_shapes:
+            params = self.get_params(**tags)
+            param_values = tf.get_default_session().run(params)
```
Comment (ryanjulian) [inline src/garage/tf/regressors/base.py]: `tf.compat.v1.get_default_session`

File: src/garage/tf/regressors/base.py [hunk 23]
```diff
@@ -36,23 +40,97 @@ def predict(self, xs):
...
+            self._cached_param_shapes[tag_tuple] = [
+                val.shape for val in param_values
+            ]
+        return self._cached_param_shapes[tag_tuple]
+
+    def get_param_values(self, **tags):
+        """Get the list of values for the parameters."""
+        params = self.get_params(**tags)
+        param_values = tf.get_default_session().run(params)
```
Comment (ryanjulian) [inline src/garage/tf/regressors/base.py]: `tf.compat.v1.get_default_session`

File: src/garage/np/policies/base.py [hunk 24]
```diff
@@ -1,83 +1,101 @@
...
-    def dist_info_sym(self, obs_var, state_info_vars):
-        """
-        Return the symbolic distribution information about the actions.
-        :param obs_var: symbolic variable for observations
-        :param state_info_vars: a dictionary whose values should contain
-         information about the state of the policy at the time it received the
-         observation
-        :return:
         """
         raise NotImplementedError
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: please use `pass`, not `NotImplementedError`

File: src/garage/np/policies/base.py [hunk 25]
```diff
@@ -1,83 +1,101 @@
...
-         information about the state of the policy at the time it received the
-         observation
-        :return:
+
+        Args:
+            obs_var (np.ndarray): observation values
+            state_info_vars (dict): a dictionary whose values should contain
+            information about the state of the policy at the time it received
+            the observation
         """
```
Comment (ryanjulian) [inline src/garage/np/policies/base.py]: `pass`

File: src/garage/tf/policies/base.py [hunk 26]
```diff
@@ -1,108 +1,274 @@
...
+    @abc.abstractmethod
     def get_action(self, observation):
-        raise NotImplementedError
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
+        Returns:
+            (np.ndarray): Action sampled from the policy.
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: no () around the type

File: src/garage/tf/policies/base.py [hunk 27]
```diff
@@ -1,108 +1,274 @@
...
+    @abc.abstractmethod
     def get_actions(self, observations):
-        raise NotImplementedError
+        """
+        Get action sampled from the policy.
+
+        Args:
+            observations (list[np.ndarray]): Observations from the environment.
+        Returns:
+            (np.ndarray): Actions sampled from the policy.
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: no () around the type

File: src/garage/tf/policies/base.py [hunk 28]
```diff
@@ -1,108 +1,274 @@
...
+        return self._cached_param_shapes[tag_tuple]
+
+    def get_param_values(self, **tags):
+        """
+        Get param values.
+
+        Args:
+            tags (dict): A map of parameters for which the values are required.
+        Returns:
+            param_values (np.ndarray): Values of the parameters evaluated in
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: no names on return values

File: src/garage/tf/policies/base.py [hunk 29]
```diff
@@ -1,108 +1,274 @@
...
 class StochasticPolicy(Policy):
+    """StochasticPolicy."""
+
     @property
     def distribution(self):
-        """
-        :rtype Distribution
-        """
+        """Distribution."""
         raise NotImplementedError
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: please use `pass`

File: src/garage/tf/policies/base.py [hunk 30]
```diff
@@ -1,108 +1,274 @@
...
+        return new_dict
+
+    def __setstate__(self, state):
+        """Object.__setstate__."""
+        self._cached_params = {}
+        self.__dict__.update(state)
 
 
 class StochasticPolicy(Policy):
+    """StochasticPolicy."""
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: this class needs to be documented too

File: src/garage/tf/policies/base.py [hunk 31]
```diff
@@ -1,108 +1,274 @@
...
-        :param state_info_vars: a dictionary whose values should contain
-         information about the state of the policy at the time it received the
-         observation
-        :return:
+
+        Args:
+            obs_var (tf.Tensor): observation values
+            state_info_vars (dict): a dictionary whose values should contain
+                information about the state of the policy at the time it
+                received the observation
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: please use `pass`

Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: this needs a return type.

File: src/garage/tf/policies/base.py [hunk 32]
```diff
@@ -1,108 +1,274 @@
...
+                information about the state of the policy at the time it
+                received the observation.
+            name (str): Name of the symbolic graph.
         """
-        raise NotImplementedError
+        pass
 
     def dist_info(self, obs, state_infos):
         """
+        Distribution info.
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: perhaps you can give an example?

File: src/garage/tf/policies/base.py [hunk 33]
```diff
@@ -1,108 +1,274 @@
...
-        """
-        :rtype Distribution
-        """
+        """Distribution."""
         raise NotImplementedError
 
+    @abc.abstractmethod
     def dist_info_sym(self, obs_var, state_info_vars, name='dist_info_sym'):
         """
+        Symbolic graph of the distribution.
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: needs a return value

File: src/garage/tf/policies/base.py [hunk 34]
```diff
@@ -1,108 +1,274 @@
...
     @property
     def distribution(self):
-        """
-        :rtype Distribution
-        """
+        """Distribution."""
         raise NotImplementedError
 
+    @abc.abstractmethod
     def dist_info_sym(self, obs_var, state_info_vars, name='dist_info_sym'):
```
Comment (ryanjulian) [inline src/garage/tf/policies/base.py]: i believe this returns the symbolic graph of the distribution parameters.

File: src/garage/tf/q_functions/base.py [hunk 35]
```diff
@@ -1,32 +1,61 @@
...
+
+        It should only copy the configuration of the Q-function,
+        not the parameters.
+
+        Args:
+            name (str): Name of the newly created q-function.
+        """
+        pass
+
+    def get_trainable_vars(self):
```
Comment (ryanjulian) [inline src/garage/tf/q_functions/base.py]: these all need types

**Comment by ryanjulian** [comment]
I think you are in for a nasty surprise now that `master` checks all pylint rules.

Most of the ones in `src/` are probably justified and easy to fix, but immediately we will probably need a different config for `tests/` vs everything else.

**Comment by nish21** [comment]
Well, in a way that's nice because I won't miss anything now. But the thought of updating docstrings of 60 files is already giving me nightmares.

**Comment by ryanjulian** [comment]
IMO fixing the docstrings is pretty easy since they are (almost) all already there and probably just need format tweaks.

I think that @gitanshu can pop in here and help you apply a different .pylintrc to the directory `tests/garage`. Everything else in `tests/` is mostly a fixture or benchmark and should probably follow the rules.

Immediate rules which don't apply to unit tests:
* Anything requiring docstrings
* `attribute-defined-outside-init`

Here's my suggested way forward
* Clean up docstrings and obvious errors in `src/`
* Get different .pylintrc running for `tests/garage` which excludes the above
* Clean up obvious errors in tests
* Come back with list of remaining violations and we can decide which ones should be disabled {everywhere, just tests, just code}

**Pull request merged**